snuba/split.py

```diff
@@ -9,86 +9,155 @@
+            # and set offset=0 so we can then trim them ourselves.
+            body['offset'] = 0
+            body['limit'] = limit - total_results + remaining_offset
+            result, status = query_func(*args, **kwargs)
+
+            # If something failed, discard all progress and just return that
+            if status != 200:
+                overall_result = result
+                break
+
+            if overall_result is None:
+                overall_result = result
+            else:
+                overall_result['data'].extend(result['data'])
+
+            if remaining_offset > 0 and len(overall_result['data']) > 0:
+                to_trim = min(remaining_offset, len(overall_result['data']))
+                overall_result['data'] = overall_result['data'][to_trim:]
+                remaining_offset -= to_trim
+
+            total_results = len(overall_result['data'])
+
+            if total_results < limit:
+                if len(result['data']) == 0:
+                    # If we got nothing from the last query, expand the range by a static factor
+                    split_step = split_step * STEP_GROWTH
+                else:
+                    # If we got some results but not all of them, estimate how big the time
+                    # range should be for the next query based on how many results we got for
+                    # our last query and its time range, and how many we have left to fetch.
+                    remaining = limit - total_results
+                    split_step = split_step * math.ceil(remaining / float(len(result['data'])))
+
+                # Set the start and end of the next query based on the new range.
+                split_end = split_start
+                try:
+                    split_start = max(split_end - timedelta(seconds=split_step), from_date)
+                except OverflowError:
+                    split_start = from_date
+
+        return overall_result, status
+
+    def col_split(*args, **kwargs):
+        """
+        Split query in 2 steps if a large number of columns is being selected.
+            - First query only selects event_id and project_id.
+            - Second query selects all fields for only those events.
+            - If the orderby term is based on timestamp also update the date range
+        """
+        body = args[0]
+        orderby = util.to_list(body.get('orderby'))
+        is_timestamp_ordered = orderby and orderby[0] in ('timestamp', '-timestamp')
+
+        min_cols = ['project_id', 'event_id']
+        if is_timestamp_ordered:
+            min_cols.append('timestamp')
+
+        minimal_query = dict(body, **{'selected_columns': min_cols})
+
+        result, status = query_func(*((minimal_query, ) + args[1:]), **kwargs)
+
+        # If something failed, just return
+        if status != 200:
+            return result, status
+
+        conditions = body.get('conditions', [])
+
+        if result:
```

Comment:
What's this conditional for? Should this be `result['data']`?

The `result` is always going to contain a bunch of extra stuff in addition to the result rows, so this will always be `True`, since `parse_and_run_query` (which is the only really valid de facto value of `query_func`) will return the value returned by `raw_query`: https://github.com/getsentry/snuba/blob/cf3e0ff076fa22bf03c223e648336a32f488f3f0/snuba/api.py#L328-L330
https://github.com/getsentry/snuba/blob/cf3e0ff076fa22bf03c223e648336a32f488f3f0/snuba/util.py#L366-L536

Reply:
Yeah, result['data'] is what it's supposed to be. If there are no results we just go ahead and do the original query.